[1단계 - 장기] 레오(김준영) 미션 제출합니다

kjyyjk opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
https://prolog.techcourse.co.kr/studylogs/4072


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [x] 2
- [ ] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
기능 구현을 하느라 정신이 없어 이미 습관이 들은 객체지향 샣활체조 원칙 외에는 잘 고려하지 못했기 때문입니다.
  
## 어떤 부분에 집중하여 리뷰해야 할까요?
안녕하세요 이프~ 7기 레오입니다.
리뷰 잘부탁드립니다.😊

이번 미션은 시간이 많이 부족하도 했고 제 눈에 아직 리팩토링할 수 있는 것들이 많이 보여서 아쉬운 점이 많은 거 같아요.
특히 기물이 이동 가능한 모든 경우를 2차원 배열로 하드 코딩하다 보니 클린 코드와는 거리가 멀어졌네요.😂

추가로 이프에게 여쭤보고 싶은 두가지가 있습니다.
### 상속 vs 조합(전략패턴)
장기에는 이동 방식이 각기 다른 다양한 종류의 기물이 존재합니다.
초반 설계에서 페어와 이를 상속으로 구현할 것인지, 전략패턴으로 구현할 것인 지에 대해 많은 고민을 했어요.

상속을 사용하면 중복 코드를 없애고 다형성을 통해 모든 기물을 Piece라는 상위 클래스로 다룰 수 있다는 장점이 있었습니다.
다만 하위 기물 클래스들이 컴파일 시점에 상위 클래스에 강결합된다는 문제가 있었어요. 
반면 전략 패턴을 사용하면 런타임 시점에 동적으로 기물의 움직임을 변경할 수 있다는 장점이 있었습니다.
하지만 다형성의 이점을 누리지 못한다는 단점이 있었는데요.

두 방법 모두 명확한 장단점을 가지고 있기 때문에 트레이드 오프가 존재한다고 생각하여 무엇을 사용할까 고민했습니다.
그리고 세가지를 근거로 이번 미션에서는 상속을 택해 구현했습니다.
1. 장기에서 기물의 움직임이 변경될 일이 있을까 생각해보았을 때 그렇지 않다라는 답변이 나왔습니다.
2. 여러 타입의 기물들을 보관하기 위해서는 다형성이 필요하다고 보았습니다.
3. 상속의 단점이 강결합이라고는 하지만 아직 그 문제를 몸소 경험해 본 적이 없습니다.

그렇게 상속을 통해서 구현을 하는데 근거 1번과는 다르게 졸/병은 진영에 따라 반대로 움직일 수 있어야 했습니다.
(초나라는 위로, 한나라는 아래로)
또한 2단계에 궁성을 구현하는 경우를 생각했을 때 기물들이 궁성 내에 들어서면 움직임이 달라지는 경우가 존재했는데요.
우선 졸병 건은 하드 코딩으로 구현했지만, 확장성까지 고려한다면 기물의 움직임 전략을 갈아 끼울 수 있는 전략 패턴을 사용하는 것이 좋지 않을까하는 생각이 들었습니다.

그래도 현재 요구사항에는 궁성을 고려할 필요가 없으니 확장이 아닌 현재에 적합한 설계를 찾는 것이 좋을까요?

이러한 고민 및 선택 과정에 피드백 해주실 만한 부분이 있다면 피드백 부탁드립니다.😀

### **instanceof**
지난 블랙잭 미션을 통해 외부에서 instance of를 사용해 하위 타입을 추론하는 행위가 캡슐화를 저해한다는 것을 학습했습니다. 이번 미션에서도 포가 포를 뛰어넘지 못하는 경우, 포가 포를 잡지 못하는 경우 등 Piece 타입 인스턴스의 하위 타입을 알아야 하는 경우가 존재했습니다. 평소였으면 Piece에 boolean을 반환하는 isCanon() 추상 메서드를 사용했을테지만, 이번에는 다른 방식으로 접근해보았습니다.

외부에서 instanceof를 사용하는 것이 캡슐화를 저해하는 것이라면, 내부에서 instanceof를 사용하면 되지 않을까 생각했고 다음과 같이 각 하위 클래스 내부에서 isSameType()을 구현해주었습니다.
```
// Canon 클래스 내부
@Override
protected boolean isSameType(Piece other) {
    return other instanceof Canon;
}
```
그러나 다시 생각해보니 `외부`라는 것이 단순히 외부의 클래스를 의미하는 것이 아니라 ```other 인스턴스 자신이 아닌 모든 것```을 의미하는 것이 아닌가 하는 생각도 들었습니다.
어디서든간에 instanceof로 other의 하위 타입을 추론하는 것은 캡슐화를 저해시키는 문제일까요?🤔

감사합니다!

kjyyjk committed:
docs: 기능 목록 추가

horizonpioneer committed:
feat: 기물의 팀을 확인하는 기능 추가

horizonpioneer committed:
feat: 졸/병 이동 여부 판단 기능 추가

kjyyjk committed:
feat: 상 이동 여부 판단 기능 추가

- 검증 로직
1. 시작 지점과 목적지를 통해 경로를 계산하고, 상이 해당 목적지로 이동할 수 있는지 검증한다.
2. 경로에 다른 기물이 있는 지 검증한다.
3. 목적지에 아군이 있는 지 검증한다.

horizonpioneer committed:
refactor: Position 클래스에 equals 재정의

horizonpioneer committed:
refactor: Piece의 자식 클래스의 공통 로직을 상속받는 방식으로 수정

- validateMovable의 내부애서 메서드나 호출 순서가 동일해서 상위 클래스에서 구현한다.
- validateMovable 내에서 차이가 발생하는 부분은 각 하위 클래스에서 구현한다.

horizonpioneer committed:
feat: 마 이동 여부 판단 기능 추가

kjyyjk committed:
feat: 포 이동 여부 판단 기능 추가

- 다음과 같은 제약 조건을 가진다.
포는 포를 뛰어넘을 수 없다.
이동 경로 사이에 기물이 하나만 있어야 한다.

- 뛰어넘는 기물이 포인지 확인하기 위해 Piece에 isSameType() 추상메서드를 생성했다.
하위 타입에서는 instance of로 파라미터로 넘어온 Piece의 인스턴스가 자신과 같은 타입인지 확인한다.
외부에서 사용 시 캡슐화가 깨진다는 이유로 instance of를 지양하는 추세이지만,
이는 각 객체에게 물어보는 형태로 instance of를 사용하기 때문에 사용 가능하다고 판단했다.

- 포는 일반 기물들과 다르게 이동 경로 안에 다른 기물이 있어야하므로 별도의 검증 메서드가 필요했다.
따라서 validatePath()를 추상메서드로 하고 각 하위 클래스에서 구현하게 했다.
이때 포를 제외한 일반 기물들의 구현 로직이 같기 때문에 Piece에 공통으로 구현 후 사용했다.

horizonpioneer committed:
feat: 차 이동 여부 판단 기능 추가

horizonpioneer committed:
feat: 사 이동 여부 판단 기능 추가

horizonpioneer committed:
feat: 궁 이동 여부 판단 기능 추가

kjyyjk committed:
feat: 보드 초기 세팅 설정

- 상차림 배치는 한나라 초나라 모두 왼상차림으로 고정한다.
(상마상마)

horizonpioneer committed:
refactor: Canon과 Chariot의 좌우 움직임 경우의 수 1 감소

horizonpioneer committed:
refactor: 진영에 따라 졸병이 다르게 움직이도록 수정

horizonpioneer committed:
feat: 출발 지점에 기물이 존재하지 않으면 예외 발생

horizonpioneer committed:
feat: 포가 포를 잡을 수 없도록 하는 기능 구현

horizonpioneer committed:
feat: Board에서 기물 이동 구현

horizonpioneer committed:
test: 목적지에 같은 진영의 기물이 있는 경우에 대한 테스트

kjyyjk committed:
test: 상의 정상적인 움직임 테스트

horizonpioneer committed:
refactor: team을 String에서 Enum으로 변경

horizonpioneer committed:
feat: 플레이어에게 수를 입력받는 기능 추가

horizonpioneer committed:
feat: 플레이어가 번갈아가며 수를 입력받는 기능 추가

horizonpioneer committed:
feat: 다른 진영 기물 이동 시 예외 발생

kjyyjk committed:
feat: 궁이 죽으면 게임을 종료하는 기능 추가

- board에서 잡힌 피스가 General인지 확인하고,
맞으면 GameOverException을 던진다.

horizonpioneer committed:
refactor: Soldier의 팀 비교 로직 수정

- String 비교를 Enum 비교 방식으로 수정하여, 한나라 졸병과 초나라 졸병이 다른 방식으로 이동하도록 수정

horizonpioneer committed:
feat: 기물 이동 시, 장기판 현황 출력 기능 추가

horizonpioneer committed:
feat: 보드 현황판 출력 시, 색 구분을 통해 피아식별 기능 추가

kjyyjk committed:
feat: 잘못된 좌표 입력 시 재입력 기능 추가

horizonpioneer committed:
refactor: 사용자의 행렬 입력을 Position 외부에서 검증

- 현재 기물의 이동 경로를 찾을 때까지 기물이 이동할 수 있는 모든 경로를 다 탐색하는 로직을 사용하고 있습니다.
- Position 내부에 좌표 범위 검증 로직이 있을 경우, 장기판 범위 밖을 탐색하게 되면 의도치 않은 예외가 발생합니다.
- try-catch 를 통해 예외를 무시하고 경로 탐색을 진행하려고 했으나, 예외를 잡아서 아무 행위도 하지 않는 것은 안티 패턴이라고
 생각하여 사용자의 입력값에 대한 검증만 진행했습니다.

horizonpioneer committed:
docs: 기능 목록 현황 업데이트

horizonpioneer committed:
test: Canon 움직임 테스트 추가

horizonpioneer committed:
test: Elephant 움직임 테스트 추가

horizonpioneer committed:
test: Horse 움직임 테스트 추가

horizonpioneer committed:
test: Soldier 움직임 테스트 추가

kjyyjk committed:
refactor: 안 쓰는 import 제거

kjyyjk commented on src/main/java/janggi/piece/Elephant.java:
```diff
@@ -0,0 +1,62 @@
+
+    private static final String NAME = "상";
+
+    private static final int[][] dRows = {{1, 1, 1}, {1, 1, 1}, {0, 1, 1}, {0, -1, -1}, {-1, -1, -1}, {-1, -1, -1}, {0, -1, -1}, {0, 1, 1}};
+    private static final int[][] dColumns = {{0, -1, -1}, {0, 1, 1}, {1, 1, 1}, {1, 1, 1}, {0, 1, 1}, {0, -1, -1}, {-1, -1, -1}, {-1, -1, -1}};
```
각 내부 배열은 해당 기물이 이동 가능한 경우를 나타냅니다.

내부 배열의 값들은 해당 이동 경우가 지나치는 경로를 의미합니다.

그리고 이를 행과 열을 구분해서 나타냈습니다.

sinb57 commented on src/main/java/janggi/view/InputParser.java:
```diff
@@ -0,0 +1,56 @@
+import static janggi.board.Board.ROW_SIZE;
+
+public class InputParser {
+    private static final String START_GOAL_DELIMETER = " ";
```
```suggestion
    private static final String START_GOAL_DELIMITER = " ";
```
오타네요ㅎ

sinb57 commented on src/main/java/janggi/board/Board.java:
```diff
@@ -0,0 +1,44 @@
+        if (piece == null) {
+            throw new IllegalArgumentException("[ERROR] 출발 지점에 기물이 존재하지 않습니다.");
+        }
+        if (!piece.isSameTeam(team)) {
```
`!` 연산자는 간혹 보지 못하고 지나치는 경우가 있어요! 이런 경우 부정의 의미를 지닌 메서드를 추가로 만들어서 사용하는 것이 좋아요~

kjyyjk replied:
의미를 잘 전달할 수 있을 거 같아 좋은 거 같아요!
```isDifferentTeam```을 추가했습니다.
1435e9cf026f983b07fdccb36c386e1f1f061628

sinb57 commented on src/main/java/janggi/piece/Piece.java:
```diff
@@ -0,0 +1,94 @@
+        }
+    }
+
+    @Override
+    public String toString() {
+        return getName();
+    }
```
toString 재정의는 콘솔 출력을 위함일까요? toString 은 무엇은 위한 메서드일까요?

kjyyjk replied:
제가 생각하기에 toString은 디버깅 시 사용하기 위함이며 그 객체의 정보를 잘 나타낼 수 있어야 한다고 생각합니다.
그런 관점에서 각 기물의 이름을 반환하는 것이 적합하다고 생각했었는데요.
이번에 이프가 주신 질문에 대한 답변을 생각하는 과정에서 기물 객체는 이름 외에도 팀을 상태로 가지기 때문에
```Team + 이름```을 반환하도록 재정의하는게 더 적합하다고 판단했습니다.
toString을 변경한 뒤 콘솔 출력 시에는 따로 이름만을 출력하게끔 수정했습니다.
86386fe6e93b86975f5f5062ac0ba37930b6f0da

sinb57 commented on src/main/java/janggi/view/InputParser.java:
```diff
@@ -0,0 +1,56 @@
+        }
+    }
+
+    private void validateRange(int column, int row) {
+        if (column < 0 || column >= Board.COLUMN_SIZE || row < 0 || row >= ROW_SIZE) {
+            throw new IllegalArgumentException("[ERROR] 올바르지 않은 좌표입니다.");
+        }
+    }
```
Position 이 지니면 어떨까요?

kjyyjk replied:
수정했습니다!

sinb57 commented on src/main/java/janggi/board/Position.java:
```diff
@@ -0,0 +1,39 @@
+
+import java.util.Objects;
+
+public class Position {
```
캐싱해보면 어떨까요?

kjyyjk replied:
캐싱..! 학습해보겠습니다!

kjyyjk replied:
안녕하세요 이프!

캐싱에 대해 학습해보고 간단한 확인도 진행해보았는데 궁금한 점이 있습니다.

캐싱을 통해 자주 사용되는 값들을 미리 생성해두고 필요할 때마다 꺼내 쓰면 객체 생성 비용과 메모리 낭비 문제를 해결할 수 있습니다.

장기 애플리케이션에서도 Position은 불변 객체이며 장기판 위 범위의 Position만 주로 사용되므로 Position도 캐싱함으로써 똑같은 이점을 얻을 수 있다고 생각했습니다.

실제로 캐싱을 한번 구현해보았고 다음과 같았습니다.

```
public class Position {
    private static final List<Position> CACHE = new ArrayList<>();

    static {
        for (Row row : Row.values()) {
            for (Column column : Column.values()) {
                CACHE.add(Position.valueOf(row, column));
            }
        }
    }

...

    public static Position valueOf(Row row, Column column) {
        Position position = null;
        for (Position cachedPosition : CACHE) {
            if (cachedPosition.row == row && cachedPosition.column == column) {
                position = cachedPosition; // break 하지 않아 최악의 경우 시간 측정
            }
        }
        if (position == null) {
            Position newPosition = new Position(column, row);
            CACHE.add(newPosition);
            return newPosition;
        }
        return position;
    }
}
```

위 코드에서는 캐싱된 값을 외부에서 꺼낼때마다 매번 ```CACHE```리스트를 탐색해야하는데요.
이 O(n) 탐색 비용이 과연 캐싱을 해서 얻는 이점보다 더 적을지 고민입니다.

간단하게 테스트를 해보았고, 다음과 같은 결과를 확인했습니다.
(아주 간단하게 진행한 테스트라 테스트 환경이 적절하지 않았을 수 있습니다.)
```
<속도>
캐싱x : 약 8ms
리스트 캐싱 : 약 100ms

<메모리>
캐싱x: 약 5800 KB
리스트 캐싱 : 약 520 KB
```

메모리 사용은 확실히 줄어든 것을 확인했지만, 속도는 약 92ms 정도 더 걸리는 것을 확인했습니다.
gpt에 의하면 실시간 애플리케이션에서 92ms는 사용자가 느낄 수 있을 정도라고 하니 과연 캐싱을 하는 것이 더 좋은 것일까 궁금합니다!

추가로 ```Map<Row<Column, Position>> CACHE``` 형태의 캐싱도 실험해보았고 ```List```보다 빠른 33ms를 보였으나 메모리 측정이 제대로 되지 않았습니다.
그래도 캐싱을 하지 않았을 때 보다는 메모리를 덜 차지할 거 같은데 Map을 사용하여 캐싱하는 것이 적절할까요?

sinb57 replied:
성능 비교까지 해보셨군요! 단일 원시값을 포장한 값객체가 아니라서 캐싱이 생각보다 효율적이지 않으리라 생각은 했는데, 직접 성능을 비교하셨다니 놀랍네요👍 `Map<Row, Map<Column, Position>>`는 아마 다를텐데, 이중으로 겹치는 만큼 저 조차도 선호하진 않네요.

캐싱을 제안드렸던 이유는, Position이 up/down/left 등의 이동을 할 때마다 새로운 Position을 만들기 때문이에요. 기물이 움직일 수 있는지와 이동 방향을 결정짓기 위해 Movement의 묶음인 Movements를 활용하고 있는데, 한두차례의 기물 움직임에도 Position 객체는 많이 생겨날 거예요. 게임이 진행될수록 일회성 객체가 계속 생겨날테니 메모리 관점에서 괜찮을지 고민해보길 바랬어요~

sinb57 commented on src/main/java/janggi/piece/Piece.java:
```diff
@@ -0,0 +1,94 @@
+        return this.team == team;
+    }
+
+    public void validateMovable(Map<Position, Piece> board, Position start, Position goal) {
```
`Map<Position, Piece>` 데이터를 넘겨받는 것이 아니라, Board 객체를 넘겨받아서 메시지를 던져보면 어떨가요?

kjyyjk replied:
이프의 제안에 동의합니다!

Board는 Map을 감싸는 일급 컬렉션으로, public 메서드 외 Map에 대한 외부 접근을 막고 있는데요.
Piece에게 내부 Map을 인자로 넘기는 행위는 이러한 캡슐화를 깨뜨린다고 판단했습니다.

또한 책임의 문제도 있다고 보았는데요.
기존처럼 Map을 Piece에게 넘기게 되면 Piece는 Board 내부의 자료구조가 Map인지를 알아야 합니다.
하지만 책임의 관점에서 봤을 때 Piece는 그저 경로에 기물이 있는지 없는 지를 알고 싶을 뿐,
Board의 내부가 어떻게 구현되어있는 지 알 필요가 없다고 생각했습니다. 

만약 기존의 방식을 유지한다면 Board 내부 구현이 바뀔 시 Piece도 바뀌게 된다는 문제가 발생할 수도 있겠네요!
이는 SRP와 OCP를 위반한다고 볼 수 있을 거 같습니다.

수정했습니다.
47d283921fe8e4d038941e4e6146b2c4430963ae

sinb57 commented on src/main/java/janggi/piece/Canon.java:
```diff
@@ -0,0 +1,108 @@
+
+    protected static final String NAME = "포";
+    private static final int[][] dRows = {
+            //상
+            {1}, {1, 1}, {1, 1, 1}, {1, 1, 1, 1}, {1, 1, 1, 1, 1}, {1, 1, 1, 1, 1, 1},
+            {1, 1, 1, 1, 1, 1, 1}, {1, 1, 1, 1, 1, 1, 1, 1}, {1, 1, 1, 1, 1, 1, 1, 1, 1},
+            //하
+            {-1}, {-1, -1}, {-1, -1, -1}, {-1, -1, -1, -1}, {-1, -1, -1, -1, -1}, {-1, -1, -1, -1, -1, -1},
+            {-1, -1, -1, -1, -1, -1, -1}, {-1, -1, -1, -1, -1, -1, -1, -1}, {-1, -1, -1, -1, -1, -1, -1, -1, -1},
+            //좌
+            {0}, {0, 0}, {0, 0, 0}, {0, 0, 0, 0}, {0, 0, 0, 0, 0}, {0, 0, 0, 0, 0, 0},
+            {0, 0, 0, 0, 0, 0, 0}, {0, 0, 0, 0, 0, 0, 0, 0},
+            //우
+            {0}, {0, 0}, {0, 0, 0}, {0, 0, 0, 0}, {0, 0, 0, 0, 0}, {0, 0, 0, 0, 0, 0},
+            {0, 0, 0, 0, 0, 0, 0}, {0, 0, 0, 0, 0, 0, 0, 0}};
```
음.. 관리하기 엄청 힘들겠네요. `패턴`에 집중하면 어떨까요?

kjyyjk replied:
패턴에 집중하라는 것이 어떤 의미일까요?

우선 움직임을 의미하는 ```Movement```라는 Enum 상수를 만들어 코드의 가독성과 유지보수성을 높였습니다!
이후 ```Movements```와 ```Path```를 생성해 움직임과 좌표에 관한 책임들을 분리했습니다.
Movement 도입 c344b0d95d38188b8cf0e47196ae31ee1f203e91
Movements 도입 f59386bb6b74189d20e03fe96fd4e8ae903d07c7
Path 도입 d1c595cfcf435d8b09e2a8ea71de81032b38e8f3

sinb57 replied:
책임 분리 잘 해주셨어요~ㅎ 훨씬 좋아졌네요👍 

---

> 패턴에 집중하라는 것이 어떤 의미일까요?

이동 가능한 모든 경우의 수를 지닐 필요가 없다고 생각했어요.
상하좌우 직선 상의 모든 거리를 이동할 수 있는 Canon 을 예로 들어보면, 1번의 up부터 10번의 up까지 총 4방향의 모든 경우의 수를 지니도록 하고 있는데요. 40가지의 값을 지니도록 하는 것이 아니라, 함수형 인터페이스로 `상하좌우 직선 방향`이라는 `패턴`을 지니면 되지 않나 싶었어요.

sinb57 reviewed (changes requested):
안녕하세요 리원! 만나서 반가워요ㅎ Level1 마지막 미션의 리뷰어를 맡은 이프 예요!
장기 미션 하시느라 고생 많으셨어요. 잘 구현해주셨네요ㅎ 코멘트 확인 부탁드려요~

---

> ... 그래도 현재 요구사항에는 궁성을 고려할 필요가 없으니 확장이 아닌 현재에 적합한 설계를 찾는 것이 좋을까요? 이러한 고민 및 선택 과정에 피드백 해주실 만한 부분이 있다면 피드백 부탁드립니다.😀

정리해주신 내용에 공감해요! 저 또한 현재 시점에서는 상속이 좋은 접근이라 생각해요. 기물의 움직임에 대해 다형성을 활용하여 풀어볼 수 있으니까요~ 그럼에도 확장성을 고려하여 전략 패턴을 사용해보시겠다면, 반대할 이유는 없어요ㅎ 직접 경험해보는 것이 좋으니까요~

질문 주신 내용에 답변 드리자면, `확장이 아닌 현재에 적합한 설계를 찾는 것이 좋을까요`가 반은 맞고 반은 틀렸어요. 상황은 언제나 달라지기 때문에 현재에 초점을 맞춘 설계를 하는 것은 맞으나, 그럼에도 확장을 고려하여 열린 설계를 하는 것 또한 중요해요! 이게 참 어려운데, 확장을 너무 고려하면 오버 엔지니어링이 되어 버려요. 그렇기 때문에 근 시일 내에 높은 확률로 변경이 이뤄날 것에 대해서만 우선 고려하는 것이 좋아요~ 지금음 궁성에 대한 스펙이 추가될 예정이니 미리 확장 가능한 구조로 설계하는 것이 좋겠죠? 하지만 예정은 예정일 뿐, 스펙이 바뀔 가능성을 고려해본다면 섣불리 확장 가능토록 구조를 입히는 것은 오히려 마이너스가 될 수 있어요. 실제로 현업에서 스펙이 하루아침에 바뀌는 경우가 많거든요. 때문에 변경 및 확장을 염두에 둘 뿐, 코드에 미리 반영하는 것은 자제하고 반영할 때의 변경이 적도록 하는 것이 최선이에요.

> ... 어디서든간에 instanceof로 other의 하위 타입을 추론하는 것은 캡슐화를 저해시키는 문제일까요?🤔

`isSameType` 메서드를 Canon 에서만 사용하는데, 이를 위해 모든 기물에 재정의하는 것은 과하지 않나 싶어요~

kjyyjk committed:
style: 코드 스타일 정리

- 불필요한 개행 및 임포트 제거
- 테스트 코드 스타일 통일화

kjyyjk committed:
style: 오타 수정

- delimeter -> delimiter

kjyyjk committed:
feat: 다른 팀인지 여부를 확인하는 메서드 추가

kjyyjk committed:
refactor: Piece의 toString 재정의 변경

- 기존에는 각 기물이 자신의 이름을 출력하는 것이 해당 객체의 정보를
잘 나타내는 것이라고 생각하여 이름을 반환하도록 toString을 재정의했다.
하지만 각 기물은 팀 정보도 가지고 있기 때문에 팀까지 같이 출력해줘야 한다고
판단해서 팀과 이름을 함께 반환하게끔 수정했다.

- 이에 따라 OutputView에서는 team에서 name을 받아와 출력하도록 수정했다.

kjyyjk committed:
refactor: Position 이동을 Movement Enum 상수화

- 기존에는 2차원 배열을 통해 기물의 모든 column 움직임과 row 움직임을
나타내주었다. 이는 가독성도 좋지 않기도 하고 잘못 입력하면 큰 장애를 낼 수
있기 때문에 유지보수성도 좋지 않다.
따라서 움직임을 Movement라는 Enum 클래스로 상수화한다.
이렇게 함으로써 타입 안정성을 챙길 수 있어 잘못된 움직임을 방지하고.
가독성과 유지보수성을 챙길 수 있다.

kjyyjk committed:
refactor: Position 내부에서 좌표 검증

- 기존에는 사용자가 입력한 좌표에 대해서만 외부에서 검증을 진행했으나,
이는 Position의 책임이라고 판단해 Position 생성 시 내부에서 검증을
하도록 수정했다.

- 시작점부터 목적지까지의 경로를 찾는 과정에서는 모든 경로를 다 찾아보는데
그 과정에서 경로가 보드판을 이탈하게 되면 예외가 발생한다.
이때는 예외가 발생하게 되는 경로를 건너뛰고 또 다른 경로를 고려해야하기
때문에 예외를 try-catch로 잡아 continue 해준다.

kjyyjk committed:
refactor: Piece에게 Map이 아닌 Board 자신을 넘기도록 수정

- Piece가 이동 가능 여부를 판단하는 과정에서 보드의 현황을 알아야하기 때문에
기존에는 보드 내부의 Map을 불변 컬렉션으로 만들어 인자로 넘겨주었다.
하지만 Board는 Map을 감싸고 있는 일급 컬렉션이다. Board는 내부가 어떻게 구현되어있는 지를 은닉하고 캡슐화한다.
그렇기 때문에 기존의 코드에서 내부 Map을 Piece에게 건네는 것은 이러한 캡슐화를 깨뜨리는 행위라고 판단했다.
그리고 기존 코드에서는 Piece가 Board 내부의 자료구조가 Map인 것을 알아야 하고, 이를 사용할 줄도 알아야한다.
하지만 책임의 관점에서 보았을 때 Piece는 특정 위치에 대한 정보를 알고 싶을 뿐이지,
Board 내부가 어떻게 구현되어있는 지는 관심이 없다.
결국 Board의 내부 구현이 변경되면 Piece도 변경 되어야한다는 문제가 있다.
그렇기 때문에 기존 코드는 SRP와 OCP를 위반했다고 판단했다.

- Piece에게 Board 자신을 넘겨서 외부에 공개된 인터페이스(public 메서드)를 활용해 정보를 받아오게 한다.

kjyyjk committed:
refactor: 경로 리스트를 감싼 Path 생성

- 기존에는 이동 경로에 존재하는 Position을 List로 관리했다.
그리고 중간 경로가 필요한 경우 외부에서 시작 Position과 끝 Position을 제거해주었다.
이처럼 경로를 저장하고 중간 경로만을 잘라 반환하는 등의 행위도 하나의 책임이라고 판단해
Path 클래스를 생성해 감싸주었다.

kjyyjk committed:
refactor: 움직임들을 감싼 Movements 생성

- Movements는 내부에 움직임을 나타내는 Movement들을 상태로 가진다.
인자로 Position을 받아 Movement를 적용한 경로를 반환한다.

kjyyjk committed:
refactor: 테스트명 수정

kjyyjk commented on src/main/java/janggi/piece/Piece.java:
```diff
@@ -0,0 +1,70 @@
+        validatePieceOnGoal(board, goal);
+    }
+
+    private Path makePath(Position start, Position goal) {
+        List<Movements> possibleMovements = getPossibleMovements();
+        for (Movements movements : possibleMovements) {
+            try {
+                Path path = movements.makePath(start);
+                if (path.lastEquals(goal)) {
+                    return path;
+                }
+            } catch (PositionOutOfBoardBoundsException e) {
+                continue;
+            }
+        }
+        throw new IllegalArgumentException("[ERROR] %s은/는 해당 목적지로 이동할 수 없습니다.".formatted(getName()));
+    }
```
기물이 위치한 지점부터 기물이 갈 수 있는 모든 경로를 탐색하도록 설계했기 때문에
기물이 어디 위치했느냐에 따라 특정 경로가 보드를 벗어나 Position에서 예외가 발생합니다.

makePath에서는 예외가 발생하는 경로 외에 적합한 경로만 찾으면 되기 때문에 해당 예외가 발생하면 다음 경로로 넘어가 처음부터 다시 탐색을 하도록 했습니다.

예외를 무시하는 건 좋은 예외처리 방법이 아니라고 알고 있는데, 이처럼 예외를 잡아서 continue하는 것 또한 예외를 무시하는 행위일까요?

제가 생각했을 때는 사용자의 입력에 대한 예외를 무시하는 것이 아니기도 하고.
특정 구간에 발생할 것이 예상되는 커스텀 예외에 대해 의도를 가지고 처리해주었기 때문에 괜찮다고 판단했습니다.

sinb57 replied:
좋은 고민이네요. 예외를 잡아서 무시하는 경우가 종종 존재하기 때문에, 무조건 나쁘다고 할 순 없어요. 그래서 레오 말처럼 괜찮을 수 있죠. 하지만 지금 로직에서 Exception이 발생하는 경우를 살펴보면, `getPossibleMovements`에서 받아온 다수의 Movements 목록을 일일히 계산하면서 exception이 발생한 케이스는 버리고 있어요. 단 하나의 Path를 만들기 위해 수십개의 Movements를 비교하는 것이 과연 옳은지 잘 모르겠네요. 최적화 과정이 필요해보여요.

kjyyjk replied:
사실 다른 크루가 여러 이동 패턴을 일반화하고 하나의 로직으로 구현한 것에 대해 설명을 들은 적이 있었는데요.
이걸 듣고 나서 같은 방식을 사용하고자 하니 마치 정답을 베끼는 것 같은 불편한 감정을 느껴서,
모든 이동 경로를 하드 코딩하고 전부 계산하는 방식을 더욱 고집했던 것 같습니다.😂

이프가 말씀 주신 것처럼 조금 더 패턴에 집중해보겠습니다!

sinb57 replied:
어떤 마음인지 깊이 공감해요. 저 또한 우테코 생활하며 그런 생각을 종종 가졌거든요. 그렇지만 다른 사람의 방법을 외면하며 성장의 기회를 놓치지 않았으면 좋겠어요. 우테코 생활 안에서 단기간에 폭풍성장할 수 있는 비결은 크루들과 환경이라고 생각해요. 크루들과 각자의 생각과 방법을 자유로이 주고받을 수 있다는 것은 엄청난 메리트이고, 이 과정에서 다른 이들의 장점과 좋은 방법을 빠르게 습득하는 것이 좋다고 생각해요. 다만 그저 받아들이는 것이 아니라, 스스로의 기준과 생각을 쌓아가며 체화하는 것이 중요하죠.

kjyyjk replied:
이프도 이런 경험이 있으시다니..!
저는 고집이 강한 성향이기도 하고  남에게 잘 설득되지 않는 편이라 그러한 학습 방식이 아직 익숙하지 않은 거 같아요. 하지만 우테코 환경의 가장 큰 이점은 주변 크루들에게 있다고 생각하기 때문에, 제 방식만을 고집하기보다는 인정할 건 인정하고 제 것으로 만들기 위한 노력을 해봐야겠어요!

sinb57 commented on src/main/java/janggi/piece/Piece.java:
```diff
@@ -0,0 +1,70 @@
+
+    public boolean isDifferentTeam(Team team) {
+        return this.team != team;
+    }
```
이런 경우 아래와 같이 메서드를 엮어주는 것이 좋아요. 로직에 수정이 필요할 때 한 곳을 누락해서 버그가 발생하는 경우가 종종 발생하기도 하구요. 메서드 체이닝을 통해 두 메서드가 서로 관련있다고 의미를 부여할 수도 있어요.
```suggestion
    public boolean isSameTeam(Team team) {
        return this.team == team;
    }

    public boolean isDifferentTeam(Team team) {
        return !isSameTeam(team);
    }
```

kjyyjk replied:
말씀해주신 부분과 ```Board#isPieceNotExsits```도 함께 수정했습니다.😊
730696b64398edee01b1905686e9fea3e499b683

sinb57 commented on src/main/java/janggi/board/Board.java:
```diff
@@ -0,0 +1,61 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class Board {
+    public static final int ROW_SIZE = 10;
+    public static final int COLUMN_SIZE = 9;
```
장기판의 규격에 관련한 값이기 때문에 Board에 위치한 것이죠? 검증을 위해 Position에서 사용되는 만큼, Position이 지니면 좋을 듯 한데 어떻게 생각하시나요?

kjyyjk replied:
동의합니다!
추가로 리팩토링을 진행했는데요,
기존에는 ```Position``` 내에서 column과 row를 원시 값으로 다루었습니다.

하지만 컴파일 시점에 유효하지 않은 값에 대해 검증이 되지 않는다는 점,
생성자의 인자 순서가 달라지면 버그가 발생한다는 점을 문제로 보았습니다.

따라서 ```Column```과 ```Row```라는 Enum 클래스를 만들어주었습니다.
이를 통해 유효하지 않은 값에 대한 컴파일 오류 확인, 타입으로 구분이 가능해졌으며,
행과 열에 대한 검증 로직을 각 클래스에 분리할 수 있었습니다.
38e4c47d6eab25788bebc0963f9effcf4a4bd79e

sinb57 commented on src/test/java/janggi/MovementsTest.java:
```diff
@@ -0,0 +1,29 @@
+        Path path = movements.makePath(start);
+
+        // then
+        assertThat(path).extracting("path", as(LIST))
```
오 `extracting`을 사용하셨군요! getter를 만들지 않기 위해 사용하신 듯 해요ㅎ
이 경우 `"path"`라는 필드명을 하드코딩해야 하는데요. 프로덕션 코드에서 필드명을 수정하면 테스트가 깨질텐데, 어떻게 생각하시나요?

kjyyjk replied:
저는 프로덕션 코드를 기반으로 테스트 코드를 작성하기 때문에 변경 사항이 전파되는 것은 자연스러운 현상이라고 생각합니다. 그런 의미에서 테스트가 깨지는 것을 두려워하지 않았고 말씀주신 부분에 대해 큰 고려를 하지 않았습니다.

이프에게 여쭤보고 싶은 질문들이 있습니다!

1) 아직 규모가 작은 애플리케이션만 다루다 보니 이런 생각을 한 것이 아닌가 싶습니다. 혹시 제 생각이 실무에서도 유효한가요?

2) 저는 필드명이 변경되었을 때 테스트가 깨진다는 단점이 있더라도 오로지 테스트를 위한 프로덕션 코드(getter 등)를 만드는 것 보다는 extracting이 더 나은 테스트 방식이라고 생각하는데요. 이프는 어떻게 생각하시나요?

sinb57 replied:
extracting이 나쁜 방법이라 생각하지 않아요~ getter를 만들지 않고서 할 수 있는 최선의 방법이라고 봐요. 때문에 레오가 잘 해주셨다고 생각하고, 반대로 단점도 인지했으면 하는 바람으로 질문을 드렸어요~


실무에서는 웹 애플리케이션 특성상 거의 모든 객체에 getter가 존재해요. 그러다보니 테스트코드 작성에서만 사용하더라도, 관성적으로 getter를 만들어두곤 해요.

그렇지만 테스트코드에서 getter를 항상 사용하진 않는데요. 객체의 내부를 검증할지, 아니면 행위에 집중할지 등 테스트로 무엇을 검증하냐에 따라 달라요. 내부를 검증할 수록 테스트코드가 많아지고 복잡해지며 프로덕션에 많이 의존하게 되지만, 행위에만 집중하는 경우 테스트를 단순화시키며 프로덕션에 덜 의존할 수 있거든요. 

프로덕션 코드의 변경에 테스트가 영향 받는 것은 당연한 흐름이지만, 때론 테스트 코드 때문에 프로덕션 코드 변경이 자유롭지 않으면 비용적으로 옳지 않을 때가 있어요. 그런 부정적인 경험은 훗날 일정에 쫓길 때 테스트 코드를 작성하지 않게끔 하곤 해요. 다음 2레벨에서 여러 형태의 테스트를 작성하다보면, 어떤 의미인지 공감되실 수 있을 거예요.

kjyyjk replied:
좋은 말씀 감사합니다!
잘 기억해두었다가 나중에 문제가 발생했을 때 이프가 말씀해주신 내용들을 다시 돌아보겠습니다.

sinb57 commented on src/main/java/janggi/Path.java:
```diff
@@ -0,0 +1,25 @@
+        path.removeFirst();
+        path.removeLast();
+        return Collections.unmodifiableList(path);
+    }
```
Path 객체의 `path` 리스트 변수를 수정하네요! `getIntermediatePath`를 호출한 후 `lastEquals`를 호출하면, 의도대로 동작한다고 볼 수 있을까요?

kjyyjk replied:
사실 ```getIntermediatePath```를 호출하고 난 뒤에는 해당 ```Path``` 객체를 더 이상 사용하지 않아
문제는 없었지만 제가 의도했던 바는 아니었습니다.
따라서 ```path``` 리스트를 복사해 중간 경로를 제외하고 반환하게끔 수정했습니다.
이 경우에는 방어적 복사가 이뤄졌다고 보아 ```unmodifiable```을 제거했습니다.

그리고 시작과 끝을 제외한 중간 경로를 반환해야 하기 때문에 ```path```의 사이즈가 2 이하인 경우 빈 리스트를 반환하게끔 했습니다.
60642945bd1868cbf28a3df87095461697741e49

sinb57 commented on src/main/java/janggi/piece/Piece.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    private Path makePath(Position start, Position goal) {
+        List<Movements> possibleMovements = getPossibleMovements();
```
`getPossibleMovements`를 호출할 때마다, 매번 `unmodifiableList`로 감싸진 컬렉션을 반환하는데요. `unmodifiableList`로 감싸는 행위가 매번 이뤄질 필요가 있을까요? 불필요한 자원 낭비로 보여요

kjyyjk replied:
외부에서 변경의 여지가 있어 불변 컬렉션을 반환하는 것이 좋겠다고 생각했는데요.

현재 구조에서는 움직임 로직이 내부에서든 외부에서든 변경 되면 안되기 때문에,
possibleMovements를 초기화할 때 List.of를 활용해 불변 리스트로 초기화 했습니다.\
이렇게 하면 메서드 호출마다 불필요한 자원을 낭비하지 않을 것이라고 생각했습니다.
39e596e8da09d8ddb06b0d7d052d6f0f035d7cbf

이프가 말씀주신 의도를 잘 이해한게 맞을까요?

sinb57 replied:
기물 객체가 possibleMovements 변수를 unmodifiableList 상태로 지니고 있었으면, 반환할 때마다 unmodifiableList로 감싸지 않고 그대로 반환할 수 있었어요. 불변이기 때문에 가능했다고 봤고, 이러한 부분에서 자원 낭비 등 성능 측면도 고려하길 바랬어요~

sinb57 commented on src/main/java/janggi/piece/Canon.java:
```diff
@@ -0,0 +1,96 @@
+        super(team);
+    }
+
+    @Override
+    protected void validatePath(Board board, Path path) {
+        int pieceCount = 0;
+        for (Position position : path.getIntermediatePath()) {
+            boolean isPieceNotExists = board.isPieceNotExists(position);
+            if (isPieceNotExists) {
+                continue;
+            }
+            boolean isCanon = board.isSameTypePieceExists(position, this);
+            if (isCanon) {
+                throw new IllegalArgumentException("[ERROR] 포는 포를 뛰어넘을 수 없습니다.");
+            }
+            pieceCount++;
+        }
+        if (pieceCount != 1) {
+            throw new IllegalArgumentException("[ERROR] 포는 다른 기물 1개를 넘어가야 합니다.");
+        }
+    }
```
2depth를 해소할 수 없을까요??

kjyyjk replied:
두 가지 검증 로직(포 존재 여부와 기물 개수)을 분리하고 Stream을 활용하도록 하여 수정했습니다.
2b289b1c0d20dab359983104471b496e45b95362

sinb57 commented on src/main/java/janggi/piece/Horse.java:
```diff
@@ -0,0 +1,71 @@
+    private static final String NAME = "마";
+    private static final List<Movements> possibleMovements;
+
+    static {
+        List<Movements> allMovements = new ArrayList<>();
```
possibleMovements 변수 할당을 static 에서 처리한 이유가 궁금해요~ 바로 변수에 할당하면 안되나요??

kjyyjk replied:
우선 현재 구조에서 기물의 이동 가능한 패턴들은 인스턴스 변수가 아니라 static 변수여야 한다고 생각했고.
초기화하면서 할당하려면 List.of를 사용했을텐데 List.of의 인자로 넘길 수 있는 Movements의 개수는 한정되어있기 때문에 static 블록 내에서 add를 해주는 것이 좋다고 생각했습니다.
또한 한줄씩 add 해주는 것이 더 가독성이 좋다고 생각했습니다.

하지만 최적화 리팩토링 중 선언과 함께 초기화가 더 나아보여서 수정했습니다.

kjyyjk replied:
그렇지만 unmodifiable 자원 낭비를 리팩토링 하면서 선언과 함께 초기화해주는 게 더 좋아보여서 수정했습니다.
39e596e8da09d8ddb06b0d7d052d6f0f035d7cbf

sinb57 commented on src/main/java/janggi/piece/Canon.java:
```diff
@@ -0,0 +1,96 @@
+        }
+    }
+
+    @Override
+    public boolean isGeneral() {
+        return false;
+    }
```
isGeneral 재정의는 General 에게만 의미 있는데요. 나머지 기물은 모두 false로 동일해요. 그렇다면 General이 아닌 기물들의 상위에 추상 클래스를 하나 둬서 중복 로직을 제거해보면 어떨까요?

kjyyjk replied:
수정했습니다!
3a70097a64f684017b519e07051e12fcae9fbd46

sinb57 reviewed (changes requested):
안녕하세요 레오~ 주말에도 미션 수행하느라 고생 많으셨어요!
이전 리뷰와 달리 객체 분리가 적극적으로 이뤄졌네요ㅎ
기물의 움직임에 대해 코멘트 남겼으니 확인 부탁드려요~

kjyyjk committed:
refactor: 메서드 재활용

kjyyjk committed:
refactor: Column과 Row Enum 클래스 생성

- 기존에는 Position 내부에서 row와 column을 int로 관리하고 있었다.
하지만 컴파일 시점에 유효하지 않은 값에 대해 검증이 되지 않는다는 점,
인자의 순서에 따라 행과 열이 바뀔 수 있다는 점이 문제라고 생각했다.

- 열은 Column, 행은 Row Enum 클래스로 감싼다.
이를 통해 컴파일 시점에 존재하지 않는 값에 대한 에러를 확인할 수 있고,
인자가 어떤 순서로 들어오든 간에 타입을 통해 구분할 수 있게 되었다.

- row와 column 각각이 검증 로직을 가진다.

kjyyjk committed:
refactor: isCanon과 isGeneral을 해당하는 하위 클래스에서만 재정의

- 기존에는 isCanon과 isGeneral과 같은 추상 메서드를 각 하위 클래스에서
모두 재정의했지만 사실 특정 하나의 클래스만 제외하고는 전부 false를
반환하므로 부모 클래스에서 false를 반환하게끔 구현하고
특정 클래스에서만 true를 반환하게끔 재구현해주었다.

kjyyjk committed:
refactor: 중간 경로 계산 로직 수정

- 기존 코드는 중간 경로 반환 시 원본 Path#path 리스트를 변경하고 있기 때문에 문제가 생길 수 있다.
따라서 List#subList를 통해 특정 구간의 새로운 리스트를 반환하게끔 수정했다.
이 경우에는 방어적 복사가 이뤄졌다고 보아 unmodifiable을 제거했다.

- 시작과 끝을 제외한 중간 경로를 반환해야 하기 때문에 path의 사이즈가 2 이하인 경우
중간 경로가 없다고 보아 빈 리스트를 반환하게끔 수정했다.

kjyyjk committed:
refactor: 기물의 움직임 경우의 수 초기화 로직 변경

- List.of로 초기화하여 불변 보장한다.

- 특수한 경우가 아니면 static 블록이 아닌 선언과 함께 초기화한다.

kjyyjk committed:
feat: 두 Position의 열/행 차이 계산 구현

kjyyjk committed:
feat: Movements가 향하는 방향을 판단하는 기능 구현

- 장기는 하나의 움직임 내에서 특정 방향의 반대로 향하지 않는다.
예를 들어 위로 향하던 중에 다시 아래로 향하지를 않는다.
따라서 Movements는 상하좌우 어떤 방향을 향해 움직이는지 계산하고 반환한다.

kjyyjk committed:
refactor: 기물의 이동 검증 로직 리팩토링

- 기존에는 시작 지점으로부터 기물이 갈 수 있는 모든 경우의 수를 다 고려하고
일치하는 경우만 경로로 사용했는데 이는 비효율적이라고 판단했다.
따라서 도착 지점과 시작 지점의 행과 열의 차이를 계산하여 이동 방향을
8개로 나누어 보았고(1,2,3,4사분면, x축y축 4방면) 동일한 이동 방향으로 움직이는
경우만 고려하여 경로를 계산하도록 수정했다.

- 기존에는 차와 포는 상하좌우 갈 수 있는 모든 경우의 수를 하나하나 다 초기화했다.
이것 또한 비효율적이라고 생각하여 상/하/좌/우로 무제한 움직일 수 있다는 새로운 패턴을
생성했다.

- 각 기물이 움직일 수 있는 경우의 수를 PossibleMovements로 감싸주었다.
PossibleMovements는 내부적으로 goal 방향의 움직임을 고려해 Path를 생성하고 반환한다.

kjyyjk commented on src/main/java/janggi/moving/PossibleMovements.java:
```diff
@@ -0,0 +1,97 @@
+import java.util.List;
+import java.util.function.Predicate;
+
+public class PossibleMovements {
```
기존에는 기물이 움직일 수 있는 모든 경우를 고려했습니다.

이를 최적화하기 위해 ```PossibleMovements```에서 goal로 향하는 방향을 계산한 뒤,
동일한 방향으로 움직이는 경우의 수만 고려하도록 리팩토링했습니다.

sinb57 replied:
잘 해주셨어요! 기물 객체들이 한결 가벼워졌네요ㅎ 다만 내부의 private 메서드들을 보면 중복이 많이 발생하고 있어요. 때문에 PossibleMovements가 무거워졌는데, 중복 제거에 대한 고민을 가져보셨으면 좋겠어요~

kjyyjk replied:
중복 로직을 최대한 줄여봤습니다..!
4b85ce48968a7f1cd5beb3499525dd9240c73a7a

kjyyjk committed:
refactor: 포의 경로 검증 로직 리팩토링

- 중간 경로를 변수화한 뒤, 중간 경로에 캐논이 존재하는 지와
중간 경로에 기물이 1개 존재하는 지를 따로 검증하도록 수정한다.

- for문 대신 Stream을 활용한다.

sinb57 commented on src/main/java/janggi/board/position/Row.java:
```diff
@@ -0,0 +1,69 @@
+import janggi.board.PositionOutOfBoardBoundsException;
+import java.util.Arrays;
+
+public enum Row {
+    NINE(9),
+    EIGHT(8),
+    SEVEN(7),
+    SIX(6),
+    FIVE(5),
```
열거형으로 만든 이유가 궁금해요~ 보드 판의 규격이 변하지 않다는 전제 때문에 상수 처리하는 것이 좋다고 판단하신 걸까요?

kjyyjk replied:
네!
도메인 특성 상 보드 판의 규격이 변하지 않을 것이고 8~9개 정도는 열거형으로 선언할 수 있겠다고 생각했습니다.

또한 Row와 Column이 장기 이동 로직의 핵심이기 때문에 컴파일 시점에 유효하지 않은 값을 확인할 수 있으면 좋을 거 같다고 판단하여 열거형으로 생성했습니다.

sinb57 commented on src/main/java/janggi/Application.java:
```diff
@@ -0,0 +1,117 @@
+
+    private static boolean playTurn(Board board, Team team) {
+        try {
+            String startAndGoal = inputView.readStartAndGoalPosition(team);
+            Position startPosition = parser.splitStartPosition(startAndGoal);
+            Position goalPosition = parser.splitGoalPosition(startAndGoal);
```
inputParser를 InputView에 결합해도 좋아 보여요~ InputView에서 Position을 반환받도록 하면, Application을 가볍게 할 수 있겠네요

kjyyjk replied:
초기에는 이프가 말씀해주신 구조를 생각해보았습니다.

하지만 사용자가 한번의 콘솔 입력에 출발과 도착 Position 모두를 입력할 수 있게 하고 싶었고,
```InputView#readStartAndGoalPosition```에서 두 값을 모두 반환할 수는 없으니 밖에서 파싱하도록 했습니다.

sinb57 replied:
그렇다면 InputView에서 두 값을 같이 반환받아도 좋지 않을까요? 보통 이럴 때 DTO를 활용하곤 해요.

sinb57 replied:
그렇다면 InputView에서 두 값을 같이 반환받아도 좋지 않을까요? 보통 이럴 때 DTO를 활용하곤 해요.

sinb57 commented on src/main/java/janggi/Application.java:
```diff
@@ -0,0 +1,117 @@
+        return new Board(board);
+    }
+
+    private static void initializeRedTeam(Map<Position, Piece> board) {
```
Application이 무거운데, 생성 초기화의 책임을 지닌 클래스는 따로 만드는게 어떤가요?

kjyyjk replied:
좋습니다~
```BoardInitializer```를 만든 뒤 ```BoardInitializer#makeBoard```를 통해 보드를 생성하게 수정했습니다.
7814b02d8068e4fb6e14fb8401c9bbed09c27e8c

sinb57 commented on src/main/java/janggi/Application.java:
```diff
@@ -0,0 +1,117 @@
+
+    private static void initializeRedTeam(Map<Position, Piece> board) {
+        Team team = Team.RED;
+        board.put(new Position(getColumn(0), getRow(6)), new Soldier(team));
```
`getColumn / getRow`를 처음 봤을 때 getter라는 인식이 강했어요! 그런데 들여다보니 정적팩토리 메서드 더군요. `get*`는 일반적으로 getter에 붙이기 때문에, 다른 네이밍이 좋아 보여요~ 정적팩토리 메서드의 보편적인 네이밍이나, 상황에 맞는 적절한 이름이 좋겠어요

kjyyjk replied:
valueOf가 적절하다고 판단해 수정했습니다!
d433f2cab37e07cb0c5ce01a79a7c9e6ddcf8ce6

kjyyjk replied:
Enum에 ```valueOf```라는 기본 메서드가 존재했었군요..!
정적 팩토리 메서드 네이밍 중 인자가 하나일때 주로 쓰이는 ```from```으로 수정했습니다.
8c1d384e2dcff062cd061e8e3da1fe1d64fc9427

sinb57 commented on src/main/java/janggi/board/position/Column.java:
```diff
@@ -0,0 +1,68 @@
+        throw new PositionOutOfBoardBoundsException("[ERROR] 더 이상 열을 감소할 수 없습니다.");
+    }
+
+    public int getValue() {
+        return value;
+    }
```
사용하지 않는 군요~

sinb57 commented on src/main/java/janggi/moving/Path.java:
```diff
@@ -0,0 +1,26 @@
+    }
+
+    public List<Position> getIntermediatePath() {
+        int pathSize = path.size();
+        if (pathSize > 2) {
+            return path.subList(1, pathSize - 1);
+        }
+        return new ArrayList<>();
```
매직넘버가 보이네요! 앞뒤의 return문 순서를 바꿔서 조건문을 수정해보면 아래처럼 매직넘버를 없앨 수 있어요
```suggestion
        if (path.isEmpty()) {
            return new ArrayList<>();
        }
        return path.subList(1, path.size() - 1);
```

kjyyjk replied:
기존 로직은
1. 아무것도 없는 경우
2. position이 1개만 들어있는 경우

두 가지 상황에서 빈 리스트를 반환하도록 하려는 의도였는데요!

이프가 제안한 로직은 2번 상황에서 IndexOutOfBounds 예외가 발생할 거 같아요.
물론 Path를 생성하기까지의 로직이 제대로 구현되었다면 2번 상황이 발생하지 않을텐데요.
이런 경우에는 믿음으로... 가도 될까요?

우선 매직 넘버를 static final 상수화해주었습니다.😊

kjyyjk replied:
아니면 Path 생성 시 길이 2 이상이어야 한다는 생성자 검증 로직을 추가하는 것도 방법일 거 같아요.

sinb57 replied:
그렇네요! size가 1일 때 예외가 발생하네요ㅠ 말씀주신 듯 생성자 검증 로직 추가가 좋을 듯 해요! Path 객체의 계약 조건으로 해결해도 좋아요~

sinb57 commented on src/main/java/janggi/moving/PossibleMovements.java:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    private List<Movements> filterMovements(Position start, Position goal) {
+        int columnDifference = goal.subtractColumn(start);
+        int rowDifference = goal.subtractRow(start);
+        if (columnDifference > 0) {
```
양수인지 비교하는 행위 조차도 Position에게 위임하면 어떨까요?

kjyyjk replied:
수정했지만 궁금한게 있습니다!

양수인지 비교 이후에 그 뺄셈 결과를 다른 메서드의 인자로 넘겨야하는데요.
성능상 큰 차이는 없겠지만 같은 뺄셈 로직을 두번 진행하게 됩니다.

그래도 값은 내부에서 비교하는게 더 좋을까요?

sinb57 replied:
음 생각해보니 두 값의 차이를 0과 비교하는 것이 아니라, 두 값의 대소 관계를 비교하면 되지 않을까요??

sinb57 commented on src/main/java/janggi/board/Board.java:
```diff
@@ -0,0 +1,70 @@
+        return board.get(position) != null;
+    }
+
+    public boolean isPieceNotExists(Position position) {
```
사용하지 않는 메서드는 정리 부탁드려요~

sinb57 commented on src/main/java/janggi/board/Board.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    public boolean isPieceExists(Position position) {
+        return board.get(position) != null;
```
```suggestion
        return board.containsKey(position);
```
null 비교가 아닌, 제공되는 메서드를 활용해보면 어떨까요?

sinb57 commented on src/main/java/janggi/moving/Movements.java:
```diff
@@ -0,0 +1,108 @@
+        path.add(start);
+        Position position = start;
+        Movement movement = movements.getFirst();
+        while(true) {
+            boolean canNotMove = movement.canNotMove(position);
+            if (canNotMove) {
+                return null;
+            }
+            position = movement.movePosition(position);
+            path.add(position);
+            if (position.equals(goal)) {
+                break;
+            }
+        }
```
while(true)는, 기능이 추가되거나 리팩터링을 하면서 코드가 변할 때 루프 탈출 조건이 도달 불가능한 상태가 되어 무한 루프에 빠질 가능성이 존재해요. 때문에 for 문으로 수정할 순 없는지 고민해보셨으면 좋겠어요~

kjyyjk replied:
직선으로 가다가 목적지를 만나면 종료해야하기 때문에 for문을 사용하기는 애매하다고 생각했습니다.
때문에 이프가 말씀주신 문제를 방지하기 위해 while(종료 조건)으로 수정했습니다.
621e67a03957a4922abf2630e6cb50c08502f0fd

sinb57 replied:
아 단순 반복이 아니었군요, 코드를 잘못 봤어요 🙇‍♂️

sinb57 reviewed (changes requested):
안녕하세요 레오~ 고생 많으셨어요!! 추가 코멘트 확인해주시고,
기물 간의 계층 구조를 고려하여 중복 코드를 제거하는 설계도 병행해보셔요~

kjyyjk committed:
rename: 변수명 및 메서드명 수정

kjyyjk committed:
refactor: 미사용 메서드 제거

kjyyjk committed:
refactor: null 비교가 아닌 Map api 사용

kjyyjk committed:
refactor: Path의 최소 길이 상수화

kjyyjk committed:
refactor: while문 괄호에 break문 지정

kjyyjk committed:
refactor: BoardInitializer 생성

kjyyjk committed:
rename: Column과 Row의 정적 팩토리 메서드명 수정

- 기존에는 valueOf였으나 Enum에는 열거형 상수의 이름을 인자로 받아
해당 상수를 반환하는 기본 valueOf 메서드가 존재한다.
따라서 정적 팩토리 메서드 네이밍 중 인자가 하나일 때 자주 쓰이는
from으로 수정하였다.

kjyyjk committed:
refactor: PossibleMovements의 중복 로직 제거

sinb57 reviewed (approved):
안녕하세요 레오~ 1단계 고생 많으셨어요! 2단계 진행을 위해 머지할게요
추가 코멘트 확인해보시고 2단계에서 답변 부탁드려요~
